Neos.Neos/Classes/FrontendRouting/CatchUpHook/DocumentUriPathProjectionHookForRouteCacheFlushing.php

```diff
@@ -0,0 +1,130 @@
+        private readonly ContentRepository $contentRepository,
+        private readonly RouterCachingService $routerCachingService,
+    ) {
+    }
+
+    public function onBeforeCatchUp(): void
+    {
+        // Nothing to do here
+    }
+
+    public function onBeforeEvent(EventInterface $eventInstance, EventEnvelope $eventEnvelope): void
+    {
+        match ($eventInstance::class) {
+//            NodeAggregateWasRemoved::class => $this->onBeforeNodeAggregateWasRemoved($eventInstance),
+//            NodePropertiesWereSet::class => $this->onBeforeNodePropertiesWereSet($eventInstance),
+//            NodeAggregateWasMoved::class => $this->onBeforeNodeAggregateWasMoved($eventInstance),
+            NodeAggregateWasDisabled::class => $this->onBeforeNodeAggregateWasDisabled($eventInstance),
+            default => null
+        };
+
+    }
+
+    public function onAfterEvent(EventInterface $eventInstance, EventEnvelope $eventEnvelope): void
+    {
+        match ($eventInstance::class) {
+//            NodeAggregateWasRemoved::class => $this->onAfterNodeAggregateWasRemoved($eventInstance),
+//            NodePropertiesWereSet::class => $this->onAfterNodePropertiesWereSet($eventInstance),
+//            NodeAggregateWasMoved::class => $this->onAfterNodeAggregateWasMoved($eventInstance),
+            NodeAggregateWasDisabled::class => $this->flushAllCollectedTags(),
+            default => null
+        };
+    }
+
+    public function onBeforeBatchCompleted(): void
+    {
+        // Nothing to do here
+    }
+
+    public function onAfterCatchUp(): void
+    {
+        // Nothing to do here
+    }
+
+    private function onBeforeNodeAggregateWasDisabled(NodeAggregateWasDisabled $event): void
+    {
+        if (!$this->isLiveContentStream($event->contentStreamId)) {
+            return;
+        }
+
+        foreach ($event->affectedDimensionSpacePoints as $dimensionSpacePoint) {
+            $node = $this->tryGetNode(fn() => $this->getState()->getByIdAndDimensionSpacePointHash(
+                $event->nodeAggregateId,
+                $dimensionSpacePoint->hash
+            ));
+            if ($node === null) {
+                // Probably not a document node
+                continue;
+            }
+
+            $this->collectTagsToFlush($node);
+        }
+    }
+
+    private function collectTagsToFlush(DocumentNodeInfo $node): void
+    {
+        array_push($this->tagsToFlush, ...$node->getRouteTags()->getTags());
+    }
+
+    private function flushAllCollectedTags(): void
+    {
+        $this->routerCachingService->flushCachesByTags($this->tagsToFlush);
+        $this->tagsToFlush = [];
+    }
+
+    private function getState(): DocumentUriPathFinder
+    {
+        return $this->contentRepository->projectionState(DocumentUriPathFinder::class);
+    }
+
+    private function isLiveContentStream(ContentStreamId $contentStreamId): bool
+    {
+        return $contentStreamId->equals($this->getState()->getLiveContentStreamId());
```

Comment:
we could move that method to the `DocumentUriPathFinder` IMO, but that's just a mini tweak ofc